lib/Sema/CSFix.cpp

```diff
@@ -1146,7 +1146,9 @@ bool NotCompileTimeConst::diagnose(const Solution &solution, bool asNote) const
   // Referencing an enum element directly is considered a compile-time literal.
   if (auto *d = solution.resolveLocatorToDecl(locator).getDecl()) {
     if (isa<EnumElementDecl>(d)) {
-      return true;
+      if (!d->hasParameterList()) {
```

Comment:
Please factor the checks out of the fix, as I mentioned on the precious PR - a fix shouldn’t be created if it’s not going to be diagnosed…

Reply:
Will do in a follow-up PR! I'm still figuring out how to refactor this code...

Reply:
I spent some time thinking about this and it's going to be hard to properly detect that argument is non-const because sometimes e.g. leading-dot syntax the overload is resolved long after `argument conversion` constraint has been simplified (result gets bound, then base gets bound and only after that the member is resolved). I think one way to archive it might to be add a new flag to a type variable e.g. `const required` which is going to get associated with "argument" type variable and get propagated to the base of the chain and from it to the member.

Also not that the check you added here is incorrect because `NotCompileTimeConst` would get a base expression as a locator but the leading-dot syntax chain this base belongs to might have more members than just one.

Reply:
> Also not that the check you added here is incorrect because NotCompileTimeConst would get a base expression as a locator but the leading-dot syntax chain this base belongs to might have more members than just one.
Thanks! Not sure I understand the situation you are describing here. Could you give a test case?

Reply:
Sure, something like:

```
enum E {
  case a
  case b(Int)
}
```

using `.a.b(42)` is going to get accepted because the the base is just `.a` and `.b(42)` is not checked.

Reply:
For this following case:
```
$cat /tmp/t.swift
enum E {
  case a
  case b(Int)
}
func foo(_ c: _const E) {

}
foo(.a.b(3))
foo(E.a.b(3))
```
I got:

```
$../build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swift-frontend -typecheck /tmp/t.swift
/tmp/t.swift:8:6: error: enum case 'b' cannot be used as an instance member
foo(.a.b(3))
    ~^
    E
/tmp/t.swift:9:7: error: enum case 'b' cannot be used as an instance member
foo(E.a.b(3))
    ~~^
    E
/tmp/t.swift:9:9: error: expect a compile-time constant literal
foo(E.a.b(3))
        ^
```
They seem to be rejected.

Reply:
How about this one:

```
enum E {
  case a
  case b(Int)

  var c: E { .b(42) }
}

func test(_: _const E) {
}

test(.a.c)
```

Reply:
ah, this one seems to be wrongly accepted.

Reply:
Filed rdar://90210674